### PR TITLE
Added well_id_96 and tm10_8_tool as required columns

### PR DESCRIPTION
**Pull request by charles-cowart**
This is the simplest update, where well_id_96 and tm10_8_tool are both required columns for pre-prep files.

### CODE REVIEW

File: metapool/tests/data/runs/230207_M05314_0346_000000000-KVMGL/sample_mapping_file.tsv [hunk 1]
```diff
@@ -1,3 +1,3 @@
...
-1	sample.2	TCCATACCGGAA	GTGTGYCAGCMGCCGCGGTAA	1	A2	108379Z	122822	1331807	1317793	Echo 550	ABTX_20230208_11052_Plate_238	ABTX_20230208_ABTX_11052	ABTX_20230208_11052_Plate_238_11.17.21.RK.FH_A2	This is a description of the experiment design.	"Illumina EMP protocol 515fbc, 806r amplification of 16S rRNA V4"	GT	Illumina	UCSDMI	12/1/01	230207_M05314_0346_000000000-KVMGL_SMPL1_S1_L001	FWD:GTGYCAGCMGCCGCGGTAA; REV:GGACTACNVGGGTWTCTAAT	Sequencing by synthesis	16S rRNA	V4	UCSDMI	SOME_CENTER_PROJECT_NAME	SOME_INSTRUMENT_MODEL	230207_M05314_0346_000000000-KVMGL	NA	NA	sample.2	HT	NA	NA
\ No newline at end of file
+	Unnamed: 0	sample_name	BARCODE	PRIMER	Primer_Plate	Well_ID	TM1000_8_tool	Primer_date	MasterMix_lot	Water_Lot	Processing_robot	Sample_Plate	Project_name	Well_description	EXPERIMENT_DESIGN_DESCRIPTION	LIBRARY_CONSTRUCTION_PROTOCOL	LINKER	PLATFORM	RUN_CENTER	RUN_DATE	RUN_PREFIX	pcr_primers	sequencing_meth	target_gene	target_subfragment	center_name	center_project_name	INSTRUMENT_MODEL	runid	extraction_robot	extractionkit_lot	orig_name	plating	tm300_8_tool	tm50_8_tool	tm10_8_tool	Well_ID_96
```
Comment (wasade) [inline metapool/tests/data/runs/230207_M05314_0346_000000000-KVMGL/sample_mapping_file.tsv:1]: Can the `Unnamed: 0` column be removed?

File: metapool/tests/test_prep.py [hunk 2]
```diff
@@ -163,7 +164,7 @@ def _check_run_230207_M05314_0346_000000000_KVMGL(self, obs):
              1, 'Echo 550', 'ABTX_20230208_11052_Plate_238_11.8.21.RK.FH_A1',
              float('nan'), float('nan'), 1317793, float('nan'), 'V4', 'A1',
              'ABTX_20230208_ABTX_11052', '108379Z', 'sample.1', 'HT', 122822,
-             1331807, float('nan')],
+             1331807, float('nan'), 'Z1', float('nan')],
```
Comment (wasade) [inline metapool/tests/test_prep.py:167]: `Z1` isn't a valid 96 well coordinate. Can the upstream code error politely if an invalid coordinate is specified?

**Review by wasade — CHANGES_REQUESTED** [review]

**Comment by charles-cowart** [comment]
Hi @mmbryant23 @wasade ! Here is a PR where the two new columns are both required in the pre-prep file and appear in the output prep-info file. I just realized I should update the pre-prep file notebook as well. That update should appear shortly.

File: metapool/tests/test_prep.py (hunk 2 above)
Comment (wasade) [reply metapool/tests/test_prep.py:167]: Thanks! Doing both would be defensive

**Comment by charles-cowart** [comment]
@wasade @mmbryant23 After reviewing MacKenzie's original email, the well_id_96 column can be added to the plate_map file input of the notebook and pass through successfully without any modification to the notebook itself. The sample plate-map file still needs that column; MacKenzie if you don't mind, can you send me a range of sample values? I can use them to add meaningful examples in the sample plate-map file and finish the validation testing for them. Thanks!

**Comment by charles-cowart** [comment]
@mmbryant23 Reviewing the existing well-id code, is it safe to assume that valid values for well_id_96 will begin with a letter A through D and end with a number 1-10? e.g.: A1, A2, ...A10?

**Comment by mmbryant23** [comment]
> 

@charles-cowart - The well_id_96 column will consist of letter A through H, and end with numbers 1-12. I'll send over an example column in a few minutes. Thanks for checking!

**Review by mmbryant23 — COMMENTED** [review]
I think this looks good to me once Z1/Z2 example is addressed - thanks, Charlie!

**Comment by charles-cowart** [comment]
@mmbryant23 @wasade Ready for review!

**Comment by charles-cowart** [comment]
Actually I noticed a loss of precision on some floating points in one of the updated TSV files. Will let you know when additional fixes are ready.

**Comment by charles-cowart** [comment]
@wasade @mmbryant23 Thank you for your patience, it's now ready for review!

**Review by mmbryant23 — APPROVED** [review]

**Comment by mmbryant23** [comment]
Thanks, Charlie! I think everything looks OK on my end. Checked all the
expected values.

File: metapool/plate.py [hunk 3]
```diff
@@ -207,6 +207,23 @@ def _decompress_well(well):
...
+    row = well[0]
+    col = well[1:]
+
+    if col.isdigit():
+        col = int(col)
+        if col in range(1, 13) and row.upper() in ['A', 'B', 'C', 'D', 'E',
+                                                   'F', 'G', 'H']:
+            return (row, col)
+
+    return None
```
Comment (wasade) [inline metapool/plate.py:224]:
```suggestion
VALID_96_WELL_COLUMNS = {i for i in range(1, 13)}
VALID_96_WELL_ROWS = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'}
def _validate_well_id_96(well):
    if well in [None, '']:
        return None
    try:
        row = well[0]
        col = well[1:]
    except IndexError:
        return None
                
    try:
        col = int(col)
    except ValueError:
        return None
        
    if col not in VALID_96_WELL_COLUMNS:
        return None
        
    if row.upper() not in VALID_96_WELL_ROWS:
        return None
        
    return (row, col)
```

File: metapool/tests/test_plate.py [hunk 4]
```diff
@@ -130,6 +130,53 @@ def test_plate_position(self):
...
+               ('G', 7), ('H', 7), ('A', 8), ('B', 8), ('C', 8), ('D', 8),
+               ('E', 8), ('F', 8), ('G', 8), ('H', 8), ('A', 9), ('B', 9),
+               ('C', 9), ('D', 9), ('E', 9), ('F', 9), ('G', 9), ('H', 9),
+               ('A', 10), ('B', 10), ('C', 10), ('D', 10), ('E', 10),
+               ('F', 10), ('G', 10), ('H', 10), ('A', 11), ('B', 11),
+               ('C', 11), ('D', 11), ('E', 11), ('F', 11), ('G', 11),
+               ('H', 11), ('A', 12), ('B', 12), ('C', 12), ('D', 12),
+               ('E', 12), ('F', 12), ('G', 12), ('H', 12)]
+
+        self.assertEqual(set(obs), set(exp))
```
Comment (wasade) [inline metapool/tests/test_plate.py:178]: The current test does not directly assert behavior of an individual value, what about the following?

```suggestion
    def test_well_id_96(self):
       tests = []
       for a in 'ABCDEFGH':
           for b in range(1, 13):
               tests.append(("f{a}{b}", (a, b)))
       for a in 'IJK':
           for b in range(1, 13):
               tests.append(("f{a}{b}", None)) 
       for a in 'ABC':
           for b in [0, 13, 14]:
               tests.append(("f{a}{b}", None))      
               
       for test, exp in tests:
           obs = _validate_well_id_96(test)
           self.assertEqual(obs, exp)                    
```

**Review by wasade — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -13,13 +13,14 @@ git clone https://github.com/biocore/metagenomics_pooling_notebook.git
 Create a Python3 Conda environment in which to run the notebook:
 
 ```bash
-conda create -n pooling_nb 'python>=3.6' scipy numpy matplotlib pandas
+conda create -n metapool 'python=3.9' scikit-learn pandas numpy scipy nose pep8 flake8 matplotlib jupyter notebook 'seaborn>=0.7.1' pip openpyxl
```
Comment (antgonza) [inline README.md:16]: Can you also change the README to reflect this change?

**Review by antgonza — COMMENTED** [review]
One extra comment, extra from @wasade suggestions.

File: README.md (hunk 5 above)
Comment (charles-cowart) [reply README.md:16]: Actually this is the readme file that's being changed.

**Comment by RodolfoSalido** [comment]
the data in the well_id_96 column in the FinRisk test_data is completely wrong. How was that made?

**Comment by charles-cowart** [comment]
> the data in the well_id_96 column in the FinRisk test_data is completely wrong. How was that made?
@RodolfoSalido It was added by myself based on the parameters MacKenzie and Daniel gave. They're correct in that they conform to what was described above. I am more than happy to replace them with better values if you can provide them!

**Comment by RodolfoSalido** [comment]
I'm not sure if we need them to have accurate values, but right now multiple samples are linked to the same Well in the well_id_96 column. I was concerned that the wells were assigned programmatically incorrectly, but it doesn't sound like that is the case.  

I stumbled into this because the new column requirement is causing some backward compatibility issues. I am updating the iSeqnorm test data to resolve the issue.

**Comment by charles-cowart** [comment]
> I'm not sure if we need them to have accurate values, but right now multiple samples are linked to the same Well in the well_id_96 column. I was concerned that the wells were assigned programmatically incorrectly, but it doesn't sound like that is the case.
> 
> I stumbled into this because the new column requirement is causing some backward compatibility issues. I am updating the iSeqnorm test data to resolve the issue.

Let me know if I can help in some way!

**Pull request merged**